sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/batch/implementation/BatchAsyncContainerExecutor.java

```diff
@@ -0,0 +1,275 @@
+        final int maxOperationCount,
+        final int maxPayloadLength) {
+
+        checkNotNull(container, "expected non-null containerCore");
+        checkArgument(maxOperationCount > 0, "expected maxServerRequestOperationCount > 0, not %s", maxOperationCount);
+        checkArgument(maxPayloadLength > 0, "expected maxServerRequestBodyLength > 0, not %s", maxPayloadLength);
+
+        this.container = container;
+        this.maxOperationCount = maxOperationCount;
+        this.maxPayloadLength = maxPayloadLength;
+        this.docClientWrapper = CosmosBridgeInternal.getAsyncDocumentClient(container.getDatabase());
+        this.throttlingRetryOptions = docClientWrapper.getConnectionPolicy().getThrottlingRetryOptions();
+    }
+
+    public Mono<TransactionalBatchOperationResult<?>> addAsync(
+        final ItemBatchOperation<?> operation,
+        final RequestOptions options) {
+
+        checkNotNull(operation, "expected non-null operation");
+
+        CompletableFuture<TransactionalBatchOperationResult<?>> future = this.validateAndMaterializeOperation(operation, options)
+            .thenComposeAsync(
+                t -> this.resolvePartitionKeyRangeIdAsync(operation)
+                    .toFuture().thenCompose(
+                        (String resolvedPartitionKeyRangeId) -> {
+
+                            BatchAsyncStreamer streamer = this.getOrAddStreamerForPartitionKeyRange(resolvedPartitionKeyRangeId);
+
+                            ItemBatchOperationContext context = new ItemBatchOperationContext(
+                                resolvedPartitionKeyRangeId,
+                                BatchAsyncContainerExecutor.getRetryPolicy(this.throttlingRetryOptions));
+                            operation.attachContext(context);
+
+                            streamer.add(operation);
+
+                            return context.getOperationResultFuture();
+                        }));
+
+        return Mono.fromFuture(future);
+    }
+
+    public CompletableFuture<Boolean> validateAndMaterializeOperation(
+        final ItemBatchOperation<?> operation,
+        final RequestOptions options) {
+
+        if (options != null) {
+            checkState(options.getConsistencyLevel() == null
+                && options.getPostTriggerInclude() == null
+                && options.getPreTriggerInclude() == null
+                && options.getSessionToken() == null, "UnsupportedBulkRequestOptions");
+
+            BatchAsyncContainerExecutor.validateOperationEpk(operation, options);
+        }
+
+        return operation.materializeResource();
+    }
+
+    private Mono<PartitionKeyRangeBatchExecutionResult> executeAsync(PartitionKeyRangeServerBatchRequest request) throws Exception {
+        request.setAtomicBatch(false);
+        request.setShouldContinueOnError(true);
+
+        final Semaphore limiter = this.getOrAddLimiterForPartitionKeyRange(request.getPartitionKeyRangeId());
+        limiter.acquire();
```

Comment:
Would it be useful to measure the latency we waited on the Semaphore and iclude it in COsmosDiagnostcis?

Reply:
CosmosDiagnostic is added later in RxDocumentClientImpl in ClientRetryPolicy, infact is replaced there with a new instance. So bit difficult to keep diagnostic before. I can keep this timeout value and later update in diagnostic, will it be ok?

Reply:
Have added it. It make sense to have semaphore wait time.